kimchi/src/circuits/polynomials/foreign_field_add/witness.rs

```diff
@@ -96,24 +107,46 @@ fn compute_subadd_values<F: PrimeField>(
         }
     });
 
-    // c1 = r2 - a2 - s*b2 + q*f2
-    let carry_mi =
+    // c = [ (a1 * 2^88 + a0) + s * (b1 * 2^88 + b0) - q * (f1 * 2^88 + f0) - (r1 * 2^88 + r0) ] / 2^176
+    //  <=>
+    // c = r2 - a2 - s*b2 + q*f2
+
+    let carry_bot: F = (bottom(&left_input[LO], &left_input[MI])
+        + bottom(&right_input[LO], &right_input[MI]) * sign
+        - bottom(&foreign_modulus[LO], &foreign_modulus[MI]) * field_overflow
+        - bottom(&result[LO], &result[MI]))
+        / (two_to_limb * two_to_limb);
+
+    let carry_top: F =
         result[HI] - left_input[HI] - sign * right_hi + field_overflow * foreign_modulus[HI];
-    // c0 = r1 - a1 - s*b1 + q*f1 + 2^88*c1
-    let carry_lo = result[MI] - left_input[MI] - sign * right_input[MI]
-        + field_overflow * foreign_modulus[MI]
-        + two_to_limb * carry_mi;
-    (result, sign, field_overflow, carry_lo, carry_mi)
+
+    // Check that both ways of computing the carry value are equal
+    assert_eq!(carry_top, carry_bot);
+
+    FFAddValues {
+        left: left_input.clone(),
+        right: right_input.clone(),
+        output: result,
+        sign,
+        ovf: field_overflow,
+        carry: carry_bot,
+    }
+}
+
+// Returns the bottom composition of an element from its low and middle limbs
+fn bottom<F: PrimeField>(lo: &F, mi: &F) -> F {
+    *lo + *mi * F::from(TWO_TO_LIMB)
```

Comment:
Once updated against master you should use `F::two_to_limb()`

Also, you should rename this to `two_limb` or `compact_limb` for consistency with constraints code.

Even better, you don't need to re-implement this function here. Thanks to `ExprOps`, you can just reuse `cirguitgates::two_limb()` everywhere.

Reply:
Ha! indeed, thanks!